code/statsd/src/main/scala/com/avast/datadog4s/statsd/event/EventerImpl.scala

```diff
@@ -0,0 +1,45 @@
+package com.avast.datadog4s.statsd.event
+
+import cats.effect.Sync
+import com.avast.datadog4s.api.Tag
+import com.avast.datadog4s.api.event.{ Eventer, _ }
+import com.timgroup.statsd.{ StatsDClient, Event => SEvent }
+
+import scala.collection.immutable.Seq
+class EventerImpl[F[_]: Sync](statsDClient: StatsDClient, defaultTags: Seq[Tag]) extends Eventer[F] {
+  private[this] val F = Sync[F]
+
+  private val toPriority: Priority.Value => SEvent.Priority = {
+    case Priority.Low    => SEvent.Priority.LOW
+    case Priority.Normal => SEvent.Priority.NORMAL
+  }
+
+  private val toAlertType: AlertType.Value => SEvent.AlertType = {
+    case AlertType.Error   => SEvent.AlertType.ERROR
+    case AlertType.Warning => SEvent.AlertType.WARNING
+    case AlertType.Info    => SEvent.AlertType.INFO
+    case AlertType.Success => SEvent.AlertType.SUCCESS
+  }
+
+  override def send(
+    event: Event,
+    tags: Tag*
+  ): F[Unit] =
+    F.delay {
+      statsDClient.recordEvent(
+        SEvent
+          .builder()
+          .withTitle(event.title)
+          .withText(event.text)
+          .withDate(event.date.map(_.toEpochMilli).getOrElse[Long](-1))
+          .withHostname(event.hostname.orNull)
+          .withAggregationKey(event.aggregationKey.orNull)
+          .withPriority(toPriority(event.priority))
```

Comment:
are you sure datadog agent handles null gracefully? I haven't checked but if not, we probably could use `.foreach`

Reply:
the [source](https://github.com/DataDog/java-dogstatsd-client/blob/master/src/main/java/com/timgroup/statsd/Event.java#L9) suggests it does, I felt the `.foreach` approach wasn't amazing as it's performing a side effect inside the Builder and the returned value was being ignored, it was unclear how the event was being built.

a cleaner approach would be to use type classes to add functionality to take relevant options so it can still be composed accordingly invokes it behind the schenes. thoughts?

Reply:
had a go at an implementation.
``` scala
import java.time.Instant

import com.timgroup.statsd.Event
import com.timgroup.statsd.Event.Builder

object EventBuilderOptionality {
  implicit class EventBuilderOptionalitySyntax(eventBuilder: Builder) {
    def withDate(instant: Option[Instant]): Builder =
      instant.map(_.getEpochSecond).map(eventBuilder.withDate).getOrElse(eventBuilder)
  }
}

import EventBuilderOptionality._

Event.builder().withDate(None).build()
```

Reply:
there is nothing wrong with using foreach here - it's not an side effect

```scala
val z: Option[Z] = ???
val builder = new Builder()
  .setX(x)
  .setY(y)

z.foreach(builder.setZ(_))
builder.build
```
something like the above is completely fine

The implicit class that you proposed is clever but i think it's unnecessarily complex - please simply use the foreach

Reply:
> The implicit class that you proposed is clever but i think it's unnecessarily complex

that's fair.

when any of the `set` functions are called the builder is returned so when you call `setZ` the builder is mutated and returned but as it's within a `foreach` the returned value is lost and builder remains mutated. I guess as we are interfacing with java its what we have to deal with but it would have liked to pass the returned builder forwards e.g.

``` scala
val z: Option[Z] = ???
val builder = new Builder()
  .setX(x)
  .setY(y)


val builderWithZ = z.map(builder.setZ).getOrElse(builder)

builderWithZ.build

```

Reply:
although this implementation could end up looking ugly

Reply:
i get where you are coming from - we tried bunch of ways to tackle this at work but in the end we kinda converged to using `foreach` ... 

the builderWithZ idea is fine but you would need to do the `x.map(...).getOrElse` for each option in question ... which im not sure is worth the it